src/ripple/overlay/MessageScheduler.md

```diff
@@ -0,0 +1,350 @@
+contention.
+We should do the same for remote jobs, i.e. messages.
+
+The model for a message scheduler can be informed by our model of a job queue.
+With the job queue, callers may submit as many jobs as they want.
+Each job is a closure representing a unit of work,
+and scheduling a job means requesting an idle thread to perform that work.
+Jobs that cannot be immediately executed are queued.
+Data for a job can continue to accumulate outside of the closure in
+anticipation of the job's execution.
+A job is finished when it returns,
+and after it finishes, its thread is returned to the pool to be
+allocated for another job.
+
+Our message scheduler has some similarities.
+Senders should be able to submit as many remote jobs as they want.
+Each remote job is a closure representing a unit of work,
+and scheduling a remote job means requesting an idle peer to perform that
+work.
+Remote jobs that cannot be immediately executed should be queued.
+Data for a remote job should continue to accumulate outside of the closure in
+anticipation of the job's execution.
+A remote job is finished when it returns a response,
+and after it finishes, its peer is returned to the pool to be allocated for
+another remote job.
+
+A peer scheduler has some differences too.
+Unlike a job, a sender starting a remote job (by sending a request message)
+cannot guarantee that it will terminate (by receiving a response message).
+Thus we need timeouts to guarantee bounded termination.
+
+Unlike local jobs where threads are isomorphic,
+peers are distinguishable.
+When a sender first sends a request,
+it may not care which peer is chosen to receive it,
+but when it retries a request, it likely will,
+perhaps preferring to send it to a different peer in hopes that it will be
+more likely to respond with the desired data in a reasonable time.
+
+
+## Definitions
+
+Before diving into the design discussion,
+I want to introduce terminology to frame that discussion.
+
+- **peer**: A connected peer server. Each peer has one or more channels.
+    Peers do not necessarily have the same numbers of channels.
+- **channel**: The unit of capacity for messaging a peer.
+    A channel is either **open** (ready to send exactly one message to the
+    peer) or
+    **closed** (giving the peer time to read and respond to the last
+    message sent).
+- **message**: A message sent to or received from a peer. A message is either
+    a request, a response, or a notification.
+- **request**: A message sent to a peer that expects a response.
+- **response** A message received from a peer in response to a request.
+- **notification**: A message sent to a peer that expects no response.
+- **sender**: An object that is waiting for an open channel to send a message.
+- **receiver**: An object that is waiting for a response.
+
+
+## Interface
+
+Let us walk through the typical workflow of a request from the perspective of
+a dependent.
```

Comment:
Yes, I mean "user of `MessageScheduler`". I try to avoid "user" because I think it implies a human to most readers, e.g. a developer that wrote code using `MessageScheduler`. I guess I prefer to just add a definition for "dependent" with this explanation.

Reply:
How about "caller" instead?